pkg/controller/networkpolicy/adminnetworkpolicy.go

```diff
@@ -0,0 +1,405 @@
+	var addressGroups []*antreatypes.AddressGroup
+	for _, peer := range peers {
+		if peer.Pods != nil {
+			addressGroup := n.createAddressGroup("", &peer.Pods.PodSelector, peer.Pods.Namespaces.NamespaceSelector, nil, nil)
+			addressGroups = append(addressGroups, addressGroup)
+		} else if peer.Namespaces != nil {
+			addressGroup := n.createAddressGroup("", nil, peer.Namespaces.NamespaceSelector, nil, nil)
+			addressGroups = append(addressGroups, addressGroup)
+		}
+	}
+	return &controlplane.NetworkPolicyPeer{
+		AddressGroups: getAddressGroupNames(addressGroups),
+	}, addressGroups
+}
+
+// processClusterSubject processes AdminNetworkPolicySubject and yield Antrea AppliedToGroups.
+func (n *NetworkPolicyController) processClusterSubject(subject v1alpha1.AdminNetworkPolicySubject) []*antreatypes.AppliedToGroup {
+	var appliedToGroups []*antreatypes.AppliedToGroup
+	var atg *antreatypes.AppliedToGroup
+	if subject.Pods != nil {
+		atg = n.createAppliedToGroup("", &subject.Pods.PodSelector, &subject.Pods.NamespaceSelector, nil)
+	} else if subject.Namespaces != nil {
+		atg = n.createAppliedToGroup("", nil, subject.Namespaces, nil)
+	}
+	if atg != nil {
+		appliedToGroups = append(appliedToGroups, atg)
+	}
+	return appliedToGroups
+}
+
+func anpActionToCRDAction(action v1alpha1.AdminNetworkPolicyRuleAction) *antreacrd.RuleAction {
+	actionAllow := antreacrd.RuleActionAllow
+	actionDrop := antreacrd.RuleActionDrop
+	actionPass := antreacrd.RuleActionPass
+	mapping := map[v1alpha1.AdminNetworkPolicyRuleAction]*antreacrd.RuleAction{
+		v1alpha1.AdminNetworkPolicyRuleActionAllow: &actionAllow,
+		v1alpha1.AdminNetworkPolicyRuleActionDeny:  &actionDrop,
+		v1alpha1.AdminNetworkPolicyRuleActionPass:  &actionPass,
+	}
+	return mapping[action]
+}
+
+func banpActionToCRDAction(action v1alpha1.BaselineAdminNetworkPolicyRuleAction) *antreacrd.RuleAction {
+	actionAllow := antreacrd.RuleActionAllow
+	actionDrop := antreacrd.RuleActionDrop
+	mapping := map[v1alpha1.BaselineAdminNetworkPolicyRuleAction]*antreacrd.RuleAction{
+		v1alpha1.BaselineAdminNetworkPolicyRuleActionAllow: &actionAllow,
+		v1alpha1.BaselineAdminNetworkPolicyRuleActionDeny:  &actionDrop,
+	}
+	return mapping[action]
+}
+
+func (n *NetworkPolicyController) processAdminNetworkPolicy(anp *v1alpha1.AdminNetworkPolicy) (*antreatypes.NetworkPolicy, map[string]*antreatypes.AppliedToGroup, map[string]*antreatypes.AddressGroup) {
+	appliedToPerRule := anpHasNamespaceLabelRule(anp)
+	appliedToGroups := map[string]*antreatypes.AppliedToGroup{}
+	addressGroups := map[string]*antreatypes.AddressGroup{}
+	var rules []controlplane.NetworkPolicyRule
+
+	for idx, anpIngressRule := range anp.Spec.Ingress {
+		var services []controlplane.Service
+		if anpIngressRule.Ports != nil {
+			services = toAntreaServicesForPolicyCRD(*anpIngressRule.Ports)
+		}
+		clusterPeers, perNSLabelPeers := splitPolicyPeerByScope(anpIngressRule.From)
+		if len(clusterPeers) > 0 || len(perNSLabelPeers) == 0 {
```

Comment:
why the branch should be executed when `len(perNSLabelPeers) == 0`? Isn't the following rule will have empty `NetworkPolicyPeer` in this case?

Reply:
Good catch. The condition is checked for ACNP because of #2383. The same situation however does not apply to AdminNetworkPolicy since peer is required to be defined